types/lodash/common/util.d.ts

```diff
@@ -221,6 +213,14 @@ declare module "../index" {
         flow<A1, R1, R2, R3, R4, R5, R6>(f1: (a1: A1) => R1, f2: (a: R1) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5, f6: (a: R5) => R6): (a1: A1) => R6;
         flow<A1, R1, R2, R3, R4, R5, R6, R7>(f1: (a1: A1) => R1, f2: (a: R1) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5, f6: (a: R5) => R6, f7: (a: R6) => R7): (a1: A1) => R7;
         flow<A1, R1, R2, R3, R4, R5, R6, R7>(f1: (a1: A1) => R1, f2: (a: R1) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5, f6: (a: R5) => R6, f7: (a: R6) => R7, ...funcs: Array<Many<(a: any) => any>>): (a1: A1) => any;
+        // 0-argument first function
+        flow<R1, R2>(f1: () => R1, f2: (a: R1) => R2): () => R2;
+        flow<R1, R2, R3>(f1: () => R1, f2: (a: R1) => R2, f3: (a: R2) => R3): () => R3;
+        flow<R1, R2, R3, R4>(f1: () => R1, f2: (a: R1) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4): () => R4;
+        flow<R1, R2, R3, R4, R5>(f1: () => R1, f2: (a: R1) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5): () => R5;
+        flow<R1, R2, R3, R4, R5, R6>(f1: () => R1, f2: (a: R1) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5, f6: (a: R5) => R6): () => R6;
+        flow<R1, R2, R3, R4, R5, R6, R7>(f1: () => R1, f2: (a: R1) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5, f6: (a: R5) => R6, f7: (a: R6) => R7): () => R7;
+        flow<R1, R2, R3, R4, R5, R6, R7>(f1: () => R1, f2: (a: R1) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5, f6: (a: R5) => R6, f7: (a: R6) => R7, ...funcs: Array<Many<(a: any) => any>>): () => any;
```

Comment:
Currently the overloads are in ascending order of the n-argument first function. Perhaps we should order the overloads in descending order?

Reply:
Yeah it looks like it - we should probably start with the 4-argument version, then 3-argument, 2-argument, etc. 

But keep the any-argument version at the end. (Although I'm not sure... maybe it should be at the beginning? Otherwise it'll break if the 5th argument is optional. If we do that, though, then we should remove the 4-argument version completely, since it will never be selected.)

Whatever you do, do the same for wrappers, and same for `flowRight`.

Reply:
> Perhaps we should order the overloads in descending order?

I gave this a go but ran into problems. In hindsight I'm no longer sure if it is the right thing to do. Here's why:

``` ts
interface LoDashStatic {
    // 2-argument first function
    flow<A1, A2, R1, R2>(f1: (a1: A1, a2: A2) => R1, f2: (a: R1) => R2): (a1: A1, a2: A2) => R2;
    // 1-argument first function
    flow<A1, R1, R2>(f1: (a1: A1) => R1, f2: (a: R1) => R2): (a1: A1) => R2;
}

declare const _: LoDashStatic;

// Actual type (a1: number, a2: {}) => number
// $ExpectType (a1: number) => number
_.flow(
    (a: number) => a,
    a => a,
);
```

So I think the right order is:

```
1-argument first function
0-argument first function
2-argument first function
3-argument first function
4-argument first function
```

WDYT?

> at the beginning? Otherwise it'll break if the 5th argument is optional.

I'm trying to understand what you meant by this so I created some tests:

``` ts
    // Actual type (a1: number, a2: number, a3: number, a4: number, ...args: any[]) => number
    // $ExpectType (a1: number, a2: number, a3: number, a4: number, ...args: any[]) => number
    _.flow(
        (a: number, b: number, c: number, d: number, e: number) => a,
        val => val
    );

    // Actual type (a1: number, a2: number, a3: number, a4: number) => number
    // $ExpectType (a1: number, a2: number, a3: number, a4: number, ...args: any[]) => number
    _.flow(
        (a: number, b: number, c: number, d: number, e?: number) => a,
        val => val
    );
```

(Note the "actual types" are the results with `strictFunctionTypes` both disabled and enabled.)

Currently the "any-argument first function" overloads come after the "4-argument first function" overloads. If I switch them, the actual type for the latter test (the optional 5th param) matches the expected type. IIUC, this is why you were saying to keep the "any-argument first function" overloads at the end? If that's correct then the changes I pushed should be good. This also meant I was able to remove the "4-argument first function" overloads, as you suggested.

Once we're happy with this I'll apply the changes elsewhere.